File: src/lanlink/invite.py
"""Pairing invitations: one link that carries address, code and certificate pin.

A LanLink invite is a ``lanlink://pair`` URL. It can be shown as a QR code for a
phone camera, or copied and pasted between two computers. It carries the
certificate fingerprint, so the receiving device can pin the right identity
instead of trusting whatever answers on that address.
"""

from __future__ import annotations

from dataclasses import dataclass
from urllib.parse import parse_qs, quote, urlencode, urlparse

SCHEME = "lanlink"
ACTION = "pair"


class InvalidInvite(ValueError):
    pass


@dataclass
class Invite:
    host: str
    port: int
    code: str
    device_id: str = ""
    name: str = ""
    fingerprint: str = ""
    scheme: str = "https"

    @property
    def base_url(self) -> str:
        return f"{self.scheme}://{self.host}:{self.port}"

    def to_url(self) -> str:
        fields = {
            "host": self.host,
            "port": str(self.port),
            "code": self.code,
            "id": self.device_id,
            "name": self.name,
            "fp": self.fingerprint,
            "scheme": self.scheme,
        }
        query = urlencode({key: value for key, value in fields.items() if value}, quote_via=quote)
        return f"{SCHEME}://{ACTION}?{query}"


def parse_invite(text: str) -> Invite:
    """Accept a lanlink:// invite, or a bare host:port typed by hand."""
    candidate = (text or "").strip()
    if not candidate:
        raise InvalidInvite("Paste a LanLink invite or an address first.")

    if candidate.lower().startswith(f"{SCHEME}://"):
        parsed = urlparse(candidate)
        if parsed.netloc and parsed.netloc.lower() != ACTION:
            raise InvalidInvite("That LanLink link is not a pairing invite.")
        values = {key: value[0] for key, value in parse_qs(parsed.query).items() if value}
        host = values.get("host", "").strip()
        port_text = values.get("port", "").strip()
        if not host or not port_text.isdigit():
            raise InvalidInvite("That invite is missing the device address.")
        return Invite(
            host=host,
            port=int(port_text),
            code=values.get("code", "").strip(),
            device_id=values.get("id", "").strip(),
            name=values.get("name", "").strip(),
            fingerprint=values.get("fp", "").strip(),
            scheme=values.get("scheme", "https").strip() or "https",
        )

    scheme = "https"
    if candidate.startswith(("http://", "https://")):
        parsed = urlparse(candidate)
        scheme = parsed.scheme
        candidate = parsed.netloc
    host, _, port_text = candidate.partition(":")
    host = host.strip("/")
    if not host:
        raise InvalidInvite("That does not look like a device address.")
    try:
        port = int(port_text) if port_text else 8765
    except ValueError as error:
        raise InvalidInvite("The port must be a number.") from error
    return Invite(host=host, port=port, code="", scheme=scheme)


def qr_matrix(payload: str) -> list[list[bool]]:
    """Render the invite to a boolean matrix a UI can paint however it likes."""
    import segno

    code = segno.make(payload, error="m")
    return [[bool(cell) for cell in row] for row in code.matrix]

File: android/core/src/Invite.kt
package link.lan.core

import java.net.URLDecoder
import java.net.URLEncoder

/**
 * §5: `lanlink://pair?...` — the string behind the QR code.
 *
 * It carries the full certificate fingerprint, so the phone pins the right
 * identity rather than trusting whatever answers on that address.
 */
data class Invite(
    val host: String,
    val port: Int,
    val code: String = "",
    val deviceId: String = "",
    val name: String = "",
    val fingerprint: String = "",
    val scheme: String = "https",
) {
    val baseUrl: String get() = "$scheme://$host:$port"

    /** §4: a full 64-character fingerprint is a pin; anything shorter is a hint. */
    val hasPin: Boolean get() = fingerprint.length == 64 && fingerprint.all { it.isHexDigit() }

    val isSecure: Boolean get() = scheme == "https"

    fun toUrl(): String {
        val fields = linkedMapOf(
            "host" to host,
            "port" to port.toString(),
            "code" to code,
            "id" to deviceId,
            "name" to name,
            "fp" to fingerprint,
            "scheme" to scheme,
        ).filterValues { it.isNotEmpty() }
        val query = fields.entries.joinToString("&") { (key, value) ->
            "$key=${URLEncoder.encode(value, "UTF-8")}"
        }
        return "lanlink://pair?$query"
    }

    companion object {
        /** Accept an invite URL, or a bare `host:port` somebody typed in. */
        fun parse(text: String): Invite {
            val candidate = text.trim()
            if (candidate.isEmpty()) throw InvalidInvite("Paste a LanLink invite or an address first.")
            return if (candidate.lowercase().startsWith("lanlink://")) {
                fromUrl(candidate)
            } else {
                fromAddress(candidate)
            }
        }

        private fun fromUrl(url: String): Invite {
            val withoutScheme = url.substring("lanlink://".length)
            val action = withoutScheme.substringBefore('?').trim('/').lowercase()
            if (action.isNotEmpty() && action != "pair") {
                throw InvalidInvite("That LanLink link is not a pairing invite.")
            }
            val values = HashMap<String, String>()
            val query = withoutScheme.substringAfter('?', "")
            for (pair in query.split('&')) {
                if (pair.isEmpty()) continue
                val key = pair.substringBefore('=')
                val raw = pair.substringAfter('=', "")
                if (key.isNotEmpty()) {
                    values[key] = runCatching { URLDecoder.decode(raw, "UTF-8") }.getOrDefault(raw)
                }
            }
            val host = values["host"].orEmpty().trim()
            val port = values["port"].orEmpty().trim().toIntOrNull()
            if (host.isEmpty() || port == null) {
                throw InvalidInvite("That invite is missing the device address.")
            }
            return Invite(
                host = host,
                port = port,
                code = values["code"].orEmpty().trim(),
                deviceId = values["id"].orEmpty().trim(),
                name = values["name"].orEmpty().trim(),
                fingerprint = values["fp"].orEmpty().trim().lowercase(),
                scheme = values["scheme"].orEmpty().trim().ifEmpty { "https" },
            )
        }

        private fun fromAddress(text: String): Invite {
            var candidate = text
            var scheme = "https"
            if (candidate.startsWith("http://") || candidate.startsWith("https://")) {
                scheme = candidate.substringBefore("://")
                candidate = candidate.substringAfter("://")
            }
            candidate = candidate.trim('/')
            val host = candidate.substringBefore(':').trim()
            if (host.isEmpty()) throw InvalidInvite("That does not look like a device address.")
            val portText = candidate.substringAfter(':', "")
            val port = if (portText.isEmpty()) DEFAULT_PORT else portText.toIntOrNull()
                ?: throw InvalidInvite("The port must be a number.")
            return Invite(host = host, port = port, scheme = scheme)
        }
    }
}

class InvalidInvite(message: String) : IllegalArgumentException(message)

private fun Char.isHexDigit(): Boolean = this in '0'..'9' || this in 'a'..'f' || this in 'A'..'F'
